fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java

```diff
@@ -249,30 +251,37 @@ private void validate(Class<?> type, FHIRPathNode node, Constraint constraint) {
             }
         }
 
-        private FHIRPathResourceNode getResource(Class<?> type, FHIRPathNode node) {
-            if (!Resource.class.isAssignableFrom(type)) {
-                // the constraint came from a data type
-                return (FHIRPathResourceNode) evaluationContext.getTree().getRoot();
-            }
-
+        private FHIRPathResourceNode getResourceNode(Class<?> type, FHIRPathNode node) {
             if (node.isResourceNode()) {
-                // the current context node is a resource node
-                return (FHIRPathResourceNode) node;
+                return node.asResourceNode();
+            }
+            
+            List<FHIRPathResourceNode> resourceNodes = getResourceNodes(node);
+            
+            if (!Resource.class.isAssignableFrom(type) && isContained(resourceNodes)) {
+                return resourceNodes.get(1);
             }
 
-            // move up in the tree to find the first ancestor that is a resource node
+            return resourceNodes.get(0);
+        }
+        
+        private boolean isContained(List<FHIRPathResourceNode> resourceNodes) {
+            return resourceNodes.size() > 1 && resourceNodes.get(1).resource().is(DomainResource.class);
+        }
+        
+        private List<FHIRPathResourceNode> getResourceNodes(FHIRPathNode node) {
```

Comment:
maybe add javadoc here (despite it being private); document the expected order of the resulting list since the calling code depends on that

Reply:
Added comments/javadoc.